test/integration/network_extension_discovery_integration_test.cc

```diff
@@ -0,0 +1,707 @@
+
+    auto configuration = test::integration::filters::TestDrainerNetworkFilterConfig();
+    configuration.set_bytes_to_drain(bytes_to_drain);
+    configuration.set_is_terminal_filter(is_terminal);
+    typed_config.mutable_typed_config()->PackFrom(configuration);
+    resource.mutable_resource()->PackFrom(typed_config);
+    if (ttl) {
+      resource.mutable_ttl()->set_seconds(1);
+    }
+    response.add_resources()->PackFrom(resource);
+    if (!second_connection) {
+      ecds_stream_->sendGrpcMessage(response);
+    } else {
+      ecds2_stream_->sendGrpcMessage(response);
+    }
+  }
+
+  void sendDataVerifyResults(uint32_t bytes_drained) {
+    test_server_->waitUntilListenersReady();
+    test_server_->waitForGaugeGe("listener_manager.workers_started", 1);
+    EXPECT_EQ(test_server_->server().initManager().state(), Init::Manager::State::Initialized);
+
+    IntegrationTcpClientPtr tcp_client = makeTcpConnection(lookupPort(port_name_));
+    ASSERT_TRUE(tcp_client->write(data_));
+    FakeRawConnectionPtr fake_upstream_connection;
+    ASSERT_TRUE(fake_upstreams_[0]->waitForRawConnection(fake_upstream_connection));
+    std::string received_data;
+    ASSERT_TRUE(
+        fake_upstream_connection->waitForData(data_.size() - bytes_drained, &received_data));
+    const std::string expected_data = data_.substr(bytes_drained, std::string::npos);
+    EXPECT_EQ(expected_data, received_data);
+    tcp_client->close();
+  }
+
+  // Verify ECDS config dump data.
+  bool verifyConfigDumpData(
+      envoy::config::core::v3::TypedExtensionConfig filter_config,
+      test::integration::filters::TestDrainerNetworkFilterConfig network_filter_config) {
+    // There is no ordering. i.e, either foo or bar could be the 1st in the config dump.
+    if (filter_config.name() == "foo") {
+      EXPECT_EQ(3, network_filter_config.bytes_to_drain());
+      return true;
+    } else if (filter_config.name() == "bar") {
+      EXPECT_EQ(4, network_filter_config.bytes_to_drain());
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  // Utilities used for config dump.
+  absl::string_view request(const std::string port_key, const std::string method,
+                            const std::string endpoint, BufferingStreamDecoderPtr& response) {
+    response = IntegrationUtil::makeSingleRequest(lookupPort(port_key), method, endpoint, "",
+                                                  Http::CodecType::HTTP1, version_);
+    EXPECT_TRUE(response->complete());
+    return response->headers().getStatusValue();
+  }
+
+  absl::string_view contentType(const BufferingStreamDecoderPtr& response) {
+    const Http::HeaderEntry* entry = response->headers().ContentType();
+    if (entry == nullptr) {
+      return "(null)";
+    }
+    return entry->value().getStringView();
+  }
+
+  const uint32_t default_bytes_to_drain_{2};
+  const std::string filter_name_;
+  const std::string data_;
+  const std::string port_name_;
```

Comment:
nit: I think these can be constexpr and maybe string_view

Reply:
It could but then it would require creating a string in functions that take std::string which repeats quick a few.